File: python/216-combination-sum-III.py
from typing import List
        
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        ans = []
        def dfs(i, path, target):
            if len(path) == k:
                if target == 0: ans.append(path[:])
                return
            for j in range(i, 10):
                if target-j < 0: continue
                path.append(j)
                dfs(j+1, path, target-j)
                path.pop()
        dfs(1, [], n)
        return ans
k = 3
n = 7
result = Solution().combinationSum3(k, n)
print(result) # [[1,2,4]]

k = 3
n = 9
result = Solution().combinationSum3(k, n)
print(result) # [[1,2,6], [1,3,5], [2,3,4]]

File: cpp/216. Combination Sum III.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
public:
    vvi ans;
    void dfs(int i, vi &path, int k, int target) {
        if(path.size() == k) {
            if(target == 0) ans.push_back(path);
            return;
        }
        for(int j = i; j <= 9; j++) {
            if(target-j < 0) continue;
            path.push_back(j);
            dfs(j+1, path, k, target-j);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum3(int k, int n) {
        vi path;
        dfs(1, path, k, n);
        return ans;
    }
};

void printAns(vvi ans) {
    for(auto l: ans) {
        cout << "(";
        for(auto n: l) {
            cout << n << " ";
        }
        cout << "), ";
    }
    cout << "\n";
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; int k, n; vvi ans;

    solution = Solution();
    k = 3, n = 7;
    ans = solution.combinationSum3(k, n);
    printAns(ans);

    solution = Solution();
    k = 3, n = 9;
    ans = solution.combinationSum3(k, n);
    printAns(ans);

    solution = Solution();
    k = 4, n = 1;
    ans = solution.combinationSum3(k, n);
    printAns(ans);

    solution = Solution();
    k = 3, n = 2;
    ans = solution.combinationSum3(k, n);
    printAns(ans);
}